### ISSUE TITLE
✨ Bun

### ISSUE THREAD

**Issue by benjaminstrasser** [dagger/dagger#6591]
### What are you trying to do?

Add official support for bun in dagger.

Below I propose a initial list of potential tasks I fell would help users to use bun with dagger

- [x] Add tests for bun #6615
- [x] Update typescript SDK to detect bun runtime ##6730
- [x] Update the documentation to show how bun is working with dagger #7054 
- [x] ~Add at least one example showcasing bun compatibility with dagger~

### Why is this important to you?

Bun is a new JavaScript runtime promising to solve many of the issues found in the current Node ecosystem that gained a lot of traction in recent months. I believe it could be a beneficial addition to the Dagger ecosystem, especially since it appears to be compatible without necessitating any changes from us (as promised). This issue should serve as a central hub for discussing whether to implement Bun support and tracking the associated tasks.

**Comment by jedevc** [comment]
cc @helderco @TomChv

**Comment by TomChv** [comment]
Hey, thanks for your issue!

I asked some question on your PR, we can move the discussion on this issue if you prefer :)

**Comment by benjaminstrasser** [comment]
As you suggested I think the general discussion fits better here. So I'll move it from #6590 to this issue.

> 1 - As far as I understand, Bun can import transpiled typescript build from node runtime into its own runtime, that what you do when you `import { connect } from "../typescript/dist"`, is that true?
> 
Yes, you are correct. Bun is mostly node compatible and it seems we do not use any of the unsupported node APIs. You can check compatibility here: [Bun Node Compaitibiliy](https://bun.sh/docs/runtime/nodejs-apis). I would want to write more tests to see if all features work inside bun and verify that assumption.


> 2 - What I had in mind with runtime support is to: have a common typescript base file composed of the generated api, entrypoint, scanner and then have a specific interface implemented with each runtime with low level functions for provisioning (because its using system call and interact with `fs`, `path`...). Do you think it's possible? I was thinking about using `node.tsconfig.json` and `bun.tsconfig.json` to specify which files to include?
> 
That is a nice idea! I actually had to look into it to figure out if anybody already did something similar to this approach and found nothing. If you found anything I would love for you to share it.

However out of curiosity I implemented two quick and dirty prototypes that kind of demonstrate how to structure this. They can be found here: [Bun-Node-Test](https://github.com/benjaminstrasser/bun-node-test).

The repo contains two examples one utilizing the method of overloading tsconfig files. It works, but might be problematic as editors do not support multiple tsconfigs and as of my understanding is not really how tsconfig files should be used. Would love your input on this as well.

The second example uses a more conventional approach of basically providing a Runtime object that can be used to access runtime specific features, such as reading files etc. The main typescript code can then interact with any arbitrary runtime we choose to support. 

I just threw those two ideas together and would love some input from you on this.

> 3 - Is it possible to use an npm dependencies in Bun? Or do we need some kind of special treatment. (e.g. do `bun add @dagger.io/dagger` would work?). In that case, no need for special low level magic trick (see question 2), we could just detect the runtime in my runtime module and choose if I use `tsx`, `bun` or `deno`
> 
`bun add @dagger.io/dagger` works fine. In the test in #6590 you could replace `import { connect } from "../typescript/dist"` with `import { connect } from '@dagger.io/dagger'` and it would work after doing a `bun install @dagger.io/dagger`. The reason I did not do that is because I build the current version of the dagger SDK with node beforehand to have the current version.

> 4 - If bun can work with npm dependencies, is it possible to execute our whole existing test suit with bun? That would avoid test duplications.

`bun run --bun mocha` does not work right now. I would have to spend more time to figure out why.
`bun test` does not work as we would need to rewrite all tests to confirm to the jest/bun syntax.

> 5 - What is the bare minimal to have a project in node being compatible with bun?
>
As far as I understand we simply cannot use any node API that is not supported yet.

___

I would like your input on the following thoughts.

I think we can continue to write the typescript sdk for node as both bun and deno aim to be node compatible. However this compatibility only works in one direction. So we could not switch bun and expect it to work for node.
Simply adding tests to make sure they are working as intended seems to be a good aproach. Here we have two options.

1. Test from the perspective of a user like I have done in #6590, which is minimally invasive
2. Get the tests, we already have to run inside the bun runtime.
This approach is not very invasive and still ensures our sdk works for node, bun and potentially deno( I have not looked into deno compatibility though).

The second approach is to have a core typescript sdk, that can utilize bun and node runtime features as discussed in my response to question 2. Here a more detailed architecture would need to be developed and I would like your input on this.
The question remains what the benefits would be(speed, native support ect.) and if those outweigh the future maintenance cost .

**Comment by TomChv** [comment]
> The second example uses a more conventional approach of basically providing a Runtime object that can be used to access runtime specific features, such as reading files etc. The main typescript code can then interact with any arbitrary runtime we choose to support.

Your example is amazing! That's exactly what I had in mind with the common interface and specialised implementation!

If we can refactor our sdk architecture to make that kind of pattern possible for:
- Interaction with host (provisioning, file management..)
- GraphQL client call

Then we can support all return in a pretty much easy way.
This would I think require a little factory to abstract the runtime loading process but it's not complex at all.

> bun run --bun mocha does not work right now. I would have to spend more time to figure out why.

I'm down to help you investigating on that if you want.

> As far as I understand we simply cannot use any node API that is not supported yet.

That's a good news here, we're not using non-standard APIs. The only one that come in my mind is typescript compiler one but I think it's available in any runtime because it's installed as a dependency.

> The second approach is to have a core typescript sdk, that can utilize bun and node runtime features as discussed in my response to question 2. Here a more detailed architecture would need to be developed and I would like your input on this.

What we could do is:
  1. Check where do we have runtime usage (I started to answer above but it's what came up from my mind, I might I've miss things)
  2. Check how we can abstract implementation into interface and implement it without breaking anything (our code has some dependencies between packages etc so it might be harder than we think)
  3. Make a small PoC
  4. Check if unit tests works for all runtime (would be a great first step)
  5. Adapt the runtime module to detect the one required (might be hard because how can we know that?), I assume there's no package.json for bun / deno, but is there a file in a deno or bun project that we could use to detect the runtime
  6. Check if integration tests works for all runtimes (might be harder but we shouldn't have much work to do except running the runtime parallel, the code should/must be the same)

⚠️ What we absolutely don't want to do is to generate a different API client depending on the runtime, this would cost us so much energy to maintain.
The less things change to support runtime, the better we are so ideally, moving only runtime related code and almost copy past it between runtime except for native function without updating any other lines would be the ideal.

> The question remains what the benefits would be(speed, native support, potential web runtime, ect.) and if those outweigh the future maintenance cost .

About the benefits of it, well it's easier to handle runtime error inside the runtime environment.
If we do a node implementation and expect it to work in Deno/Bun, we might ends with specific runtime issue that we cannot fixe (it's an assumption).

There's one extra thing I'm thinking about: 
What runtime do we use by default on `dagger init`? If it's a config choice from user input, how do we make it standard because there's no such thing with python/go. 
It's an open question and I don't have an answer yet.

**Comment by benjaminstrasser** [comment]
> > bun run --bun mocha does not work right now. I would have to spend more time to figure out why.
> 
> I'm down to help you investigating on that if you want.
> 

`bun run --bun mocha` works now as can be seen in #6615

> What we could do is:
> 
> 1. Check where do we have runtime usage (I started to answer above but it's what came up from my mind, I might I've miss things)
> 2. Check how we can abstract implementation into interface and implement it without breaking anything (our code has some dependencies between packages etc so it might be harder than we think)
> 3. Make a small PoC
> 4. Check if unit tests works for all runtime (would be a great first step)
> 5. Adapt the runtime module to detect the one required (might be hard because how can we know that?), I assume there's no package.json for bun / deno, but is there a file in a deno or bun project that we could use to detect the runtime
> 6. Check if integration tests works for all runtimes (might be harder but we shouldn't have much work to do except running the runtime parallel, the code should/must be the same)
> 
That sounds like a good idea if we decide to move forward with this idea.

> > The question remains what the benefits would be(speed, native support, potential web runtime, ect.) and if those outweigh the future maintenance cost .
> 
> About the benefits of it, well it's easier to handle runtime error inside the runtime environment. If we do a node implementation and expect it to work in Deno/Bun, we might ends with specific runtime issue that we cannot fixe (it's an assumption).
> 
I agree that we would benefit from more flexibility. As for bun, it seems to be a lot of work for potential issues. I do not know if deno could run dagger if we do not do this. That is definitely something worth investigating.


---

Some additional thoughts on my side:
Bun already has a native implementation for all node APIs that we use and all unit tests run inside the bun runtime (see #6615). Here it looks unnecessary at a glance to wrap all node/runtime functionality ourselves, as bun seems to have already done so natively anyway. However this might not work as smoothly for all runtimes we aim to support, or some problems might arise in the future for bun as well, that might be easier to fix if we separate the runtime specific code.

The question for me remains, and I would love your input on this, if such problems should  be fixed by Bun and Deno not ourselves as they aim for node/npm -compatibility anyway.

I would also like to offer a discord call on this subject if your interested. I feel this would help us to align more quickly and move towards formulating concrete next steps.

**Comment by TomChv** [comment]
> bun run --bun mocha works now as can be seen in https://github.com/dagger/dagger/pull/6615 🥳

Yeaaah congratz!

> The question for me remains, and I would love your input on this, if such problems should be fixed by Bun and Deno not ourselves as they aim for node/npm -compatibility anyway.

I think there's no good answer to this, ideally it should be up to `Bun` and `Deno` but we also need to have control on our codebase.
If we hit some kind of issue due to non-Node compatibility and we just wait for a fix, we will depend on the Deno/Bun community to make our SDK compatible

> I would also like to offer a discord call on this subject if your interested. I feel this would help us to align more quickly and move towards formulating concrete next steps.

You can reach out on Discord and we can setup a date 🚀

**Comment by TomChv** [comment]
We just had an hour call with @benjaminstrasser to talk about the support of Bun/Deno in the TypeScript SDK.

Here's a summary of our discussion:
@benjaminstrasser Feel free to add any comments below to complete my summary.

## Support Deno

Bad new 😢 Deno does not support npm imports without code modification.

Benjamin told me about [Denoify](https://github.com/garronej/denoify) which is a tool that could help us port our TypeScript SDK to be Deno compatible but this requires extra build step and maybe more work to fix the module if something fails during the build.

So we're going to delay the support for Deno.
The [fluentci module still exist](https://github.com/fluent-ci-templates/deno-pipeline/tree/ef13c582102a6c7ecc266b766e6457e7e12fa847) and could be used as a workaround.

## Support Bun

Good News 🚀 Bun and Node are 100% compatible with our needs so we can keep the same codebase and codegen.
The only difference would basically be the entrypoint.

The difference between this and [fluentci](https://github.com/fluent-ci-templates/bun-pipeline/tree/0f4d26870110a63d673d9e7b27972e33ebc0a998) is that we want to make it dagger CLI compatible.
Basically, the DX doesn't change (or change as little as possible), only the runtime.

### DX changes

In order to officially integrate the Bun runtime to the TypeScript SDK, we would need to slightly change the DX of `init` so the user can choose which runtime to use with its SDK.

**Example**

```shell
dagger init --name=foo --sdk=typescript --runtime=node # (we choose node by default anyway)
```

```shell
dagger init --name=foo --sdk=typescript --runtime=bun
```

**BUT there's a way to make it without changing the CI**

We can have a workaround before integrating it to the official Dagger CLI if we do a bun runtime module and use it as the SDK.

```shell
dagger init --name=foo --sdk=github.com/<url>/bun-runtime 
```

However, we would still need to figure a way to test it inside our CI without duplicating all Typescript test, like we do with elixir, rust... 

/cc @shykes @sipsma @vito @jpadams

### Runtime

As we said, we can support Bun simply by creating a Bun runtime that will be a copy of our [Typescript](https://github.com/dagger/dagger/blob/4db1cf99bd4f7077dafea2c02a462037d9b48090/sdk/typescript/runtime/main.go) runtime but with bun setup in it.

This is a list of what we need to do:
- [ ] Add `ModuleRuntime` and `Codegen` functions
- [ ] Setup Bun container
- [ ] Bind a template (could be a copy paste of the one in our TS runtime)
- [ ] Get Dagger Codegen (It's gonna be tricky because the Bun runtime will not be part of the official Dagger bin for now, but we can probably fetch Dagger sources and compile it inside our module as a workaround)
- [ ] Execute codegen
- [ ] Setup Bun entrypoint
- [ ] Setup cache for performance optimization

Normally after doing that, Bun can be our new runtime.

/cc @vito What's your opinion on that?

### Documentation

We should add some contents in our doc to explain how to use the Bun runtime with the TypeScript SDK.

I don't think this need a panel like we had with EcmaScript, JS and TS because the code will be exactly the same.
The work is more about how to tell Dagger to use the Bun Runtime instead of the Node one in the init step (see above).

/cc @vikram-dagger 

### Tests

With our current Dagger version, #6615 is enough because we're not testing in the context of Module yet but once Module are officially supported, we will need to update Bun test to be done inside the module context.

Meaning: we should have a way to test external runtimes done by users.
Bun is a bit special because the test codes can be shared with the TypeScript one but still, I don't know how we can test it.

@sipsma @vito 

## Node version

We also talk about the bump of the Node version in our mage file.
We should test with both `18` and `20` because `16` is no longer maintained.

@benjaminstrasser will open a PR about it soon.

**Comment by vito** [comment]
@TomChv Sounds good to me overall!

> DX changes

Is there a way the TypeScript SDK could detect that you want to use the Bun runtime using some existing convention in the JS ecosystem? We've repeatedly brought up the idea of passing params to SDKs, since it seems like a natural fit with module constructors etc., but so far we haven't needed it. Just curious mostly, there's a first time for everything.

We thought we would need SDK params to configure the Go version for the Go SDK, but it turns out Go supports configuring the toolchain natively and now we respect it (#6652). I'm guessing there's some way to configure the Python version too, or maybe that's blocked on SDK configurability (cc @helderco :pray:).

Either way it seems like we wouldn't need to copy-paste the TS SDK considering the code should theoretically be the same, it just execs a different runtime. At worst seems like we could have two modules that share code. Or one generic parameterizable module that is called by two unparameterizable SDK modules. (That's if we don't want to support SDK params for some reason.)

**Comment by benjaminstrasser** [comment]
> Is there a way the TypeScript SDK could detect that you want to use the Bun runtime using some existing convention in the JS ecosystem?
>
I am not aware of any conventions.

We've repeatedly brought up the idea of passing params to SDKs, since it seems like a natural fit with module constructors etc., but so far we haven't needed it. Just curious mostly, there's a first time for everything.
> 
This would probably be an easy to implement solution. We could read the newly added `runtime`(or any other name) field from the ModuleConfig like we do with the `name` right now and decide, which runtime should be used defaulting to node. Don't know what would happen if the field is set inside a project utilizing the go sdk for example. Maybe that is a bit confusing?


> Either way it seems like we wouldn't need to copy-paste the TS SDK considering the code should theoretically be the same, it just execs a different runtime. At worst seems like we could have two modules that share code. Or one generic parameterizable module that is called by two unparameterizable SDK modules. (That's if we don't want to support SDK params for some reason.)
>
I think you are right. Two modules are also a possibility.

---
I wrote a quick PoC that essentially just changed 7 lines inside our current TS SDK runtime to now use bun to demonstrate the similarity. [Commit](https://github.com/benjaminstrasser/dagger/commit/ec700e72dcd279df1f2fbbcf41bdb3e040cfcb16)

With those changes I can init a new ts module with `dagger init` and run both examples inside the bun runtime with `dagger call`.

**Comment by TomChv** [comment]
> Is there a way the TypeScript SDK could detect that you want to use the Bun runtime using some existing convention in the JS ecosystem? We've repeatedly brought up the idea of passing params to SDKs, since it seems like a natural fit with module constructors etc., but so far we haven't needed it. Just curious mostly, there's a first time for everything.

If a `bunlock` exists. But there's still the case of: how do you init a project with Bun as runtime?

> Don't know what would happen if the field is set inside a project utilizing the go sdk for example. Maybe that is a bit confusing?

That's why I'm not 100% comfortable with that idea, if we add a `runtime` field to the CLI, we must be consistent.

> I think you are right. Two modules are also a possibility.

It's the simplest/best one for now I think.

> I wrote a quick PoC that essentially just changed 7 lines inside our current TS SDK runtime to now use bun to demonstrate the similarity. [Commit](https://github.com/benjaminstrasser/dagger/commit/ec700e72dcd279df1f2fbbcf41bdb3e040cfcb16)

That's great! Now I think we should do the same but in an external module so we can plug it to Dagger as a module.
If you want I can make on on my daggerverse repository: https://github.com/quartz-technology/daggerverse
Or you can do it on yours and I can contributes to it, I'm comfortable with both options ;)

**Comment by helderco** [comment]
> Is there a way the TypeScript SDK could detect that you want to use the Bun runtime using some existing convention in the JS ecosystem? We've repeatedly brought up the idea of passing params to SDKs, since it seems like a natural fit with module constructors etc., but so far we haven't needed it. Just curious mostly, there's a first time for everything.

Yes, I think this is still needed. I’ve talked about other ways to go about it for certain use cases in a zenith community call, and will elaborate further. It can be a more performant solution since you don’t need to do discovery, but haven’t compiled a list of use cases where it would actually be a requirement, so I don’t have that answer top of mind at the moment.

> > Don't know what would happen if the field is set inside a project utilizing the go sdk for example. Maybe that is a bit confusing?
> 
> That's why I'm not 100% comfortable with that idea, if we add a `runtime` field to the CLI, we must be consistent.

The way I discussed this previously, or how I envision this, is that a **runtime** module’s constructor would produce config that’s associated to the `"sdk"` field in `dagger.json`. I think this is easy to understand, it’s not from *any* module, it’s from the module that defines the SDK itself, and it should be static in `dagger.json`, not at runtime in the CLI.

So the config for `"sdk": "typescript"` is defined in the constructor for the [Typescript runtime module](https://github.com/dagger/dagger/tree/main/sdk/typescript/runtime) (built-in). For a custom `"sdk": "github.com/acme/runtime"`, it’s in acme/runtime’s constructor.

> We thought we would need SDK params to configure the Go version for the Go SDK, but it turns out Go supports configuring the toolchain natively and now we respect it ([#6652](https://github.com/dagger/dagger/pull/6652)).

How does that work? In `From`, a specific version is downloaded. Does that setting make Go download a different version automatically, or does it simply ignore or not complain if the user’s version is different from the one in the container?

> If a `bunlock` exists. But there's still the case of: how do you init a project with Bun as runtime?

It’s not planned in the Python SDK to support multiple runtimes, but it is to support different package managers and even Python versions. Choosing a different runtime can go as far back as choosing a different docker image in `From`, but the same is true for a different version.

The solution is to do the discovery in Go, early in the [`Base` function](https://github.com/dagger/dagger/blob/8609cc4bd91f92e33bb5746a070348bae7ed2c23/sdk/typescript/runtime/main.go#L149). So you need to just look in `modSource` and look for `bunlock`. If it’s there, then choose a different `From` and whatever else. This includes making changes to the files that come out of codegen, if necessary, to support the different runtime.

Additionally in `pyproject.toml`, for example, we can add extra configs under a custom “category”, so until we have `sdk` specific configs in `dagger.json` from the runtime module’s constructor, I can add those configs to `pyproject.toml` instead, have Go read the `.toml` file, and set things appropriately.

However, I’ll at least try to resort to discovery as much as possible to avoid requiring explicit configs when something can be detected.

**Comment by TomChv** [comment]
> So the config for "sdk": "typescript" is defined in the constructor for the [Typescript runtime module](https://github.com/dagger/dagger/tree/main/sdk/typescript/runtime?rgh-link-date=2024-02-19T13%3A15%3A32Z) (built-in). For a custom "sdk": "github.com/acme/runtime", it’s in acme/runtime’s constructor.

It already works no? I talked about it there: https://github.com/dagger/dagger/issues/6591#issuecomment-1948483701

And this would solve all our problems actually

**Comment by helderco** [comment]
> > So the config for "sdk": "typescript" is defined in the constructor for the [Typescript runtime module](https://github.com/dagger/dagger/tree/main/sdk/typescript/runtime?rgh-link-date=2024-02-19T13%3A15%3A32Z) (built-in). For a custom "sdk": "github.com/acme/runtime", it’s in acme/runtime’s constructor.
> 
> It already works no? I talked about it there: [#6591 (comment)](https://github.com/dagger/dagger/issues/6591#issuecomment-1948483701)

No, you can use a custom SDK and it can have a constructor, but it's not used as additional config in `dagger.json`, which was my point. However if there's any way to detect from the user's source files the intended runtime, I think that's preferrable.

**Comment by TomChv** [comment]
> Since a package.json only exists inside a typescript module this could be a nice middle ground. Would love your opinions on this.

This is not a bad idea! However, it's also possible that the user install dagger in the context of an existing project directory (even if I definitely do not recommend that). But in that case, we would need to update its package.json too there too.

Also what about Deno? This has no package.json afaik

**Comment by benjaminstrasser** [comment]
> > Since a package.json only exists inside a typescript module this could be a nice middle ground. Would love your opinions on this.
> 
> This is not a bad idea! However, it's also possible that the user install dagger in the context of an existing project directory (even if I definitely do not recommend that). But in that case, we would need to update its package.json too there too.
> 
I think dagger should not update the package.json, but the user can set dagger.runtime to node|bun inside the package.json. Otherwise we just try to detect the runtime on our own unless it is specified.


> Also what about Deno? This has no package.json afaik

deno also claims to support package.json. Otherwise we could simply detect the deno.json or/and also look inside the deno.json for a custom dagger field.

---
I would suggest this workflow and would love your input on this.

someone runs `dagger init --sdk typescript` -> we generate a typescript project the way we do now potentially adding a new dagger.runtime field inside the package.json 

whenever some uses `dagger call` we check the following conditions:
```ts
function decideRuntime(workdir) {
  packageJson = getPackageJson(workdir);
  if (packageJson.exists()) {
    if (
      packageJson.dagger.runtime == "bun" ||
      packageJson.dagger.runtime == "node"
    ) {
      return packageJson.dagger.runtime;
    }
  }

  if (PackageLockJsonExists(workdir)) {
    return "node";
  }

  if (BunLockJsonExists(workdir)) {
    return "bun";
  }

  return "node";
}
```

This way we can allow the user to specify the runtime inside the `package.json` if they wish to specify it and simply try to detect it if nothing is specified. If we figure out more ways to detect the wanted runtime we can simply add them to our detection code, which is only triggered if no runtime is specified inside the `package.json`

This way we can easily detect the runtime once a user runs `bun install` or `npm install`, while still allowing for specific configuration.

**Comment by TomChv** [comment]
> This way we can allow the user to specify the runtime inside the package.json if they wish to specify it and simply try to detect it if nothing is specified. If we figure out more ways to detect the wanted runtime we can simply add them to our detection code, which is only triggered if no runtime is specified inside the package.json

> This way we can easily detect the runtime once a user runs bun install or npm install, while still allowing for specific configuration.

This looks good to me! It also simpler to maintain than having a special bun runtime module, I like the idea! You found  a nice workaround here.

What's your opinion @helderco ?

**Comment by helderco** [comment]
> What's your opinion @helderco ?

It's the same as what I said above in https://github.com/dagger/dagger/issues/6591#issuecomment-1952429481:

> Additionally in `pyproject.toml`, for example, we can add extra configs under a custom “category”, so until we have `sdk` specific configs in `dagger.json` from the runtime module’s constructor, I can add those configs to `pyproject.toml` instead, have Go read the `.toml` file, and set things appropriately.

I was drawing a parallel between `pyproject.toml` and `package.json`. But I also added this:

> However, I’ll at least try to resort to discovery as much as possible to avoid requiring explicit configs when something can be detected.

**Comment by benjaminstrasser** [comment]
> This looks good to me! It also simpler to maintain than having a special bun runtime module, I like the idea! You found a nice workaround here.
>

I think it was heavily inspired by @helderco's comments ;D

> 
> Let me know if you try that in a PR :D
> 

I will start working on this as soon as I can.

**Comment by TomChv** [comment]
> IMO updating the docs and providing a simple example are the next steps.

Yep definitely, I was thinking the same.

> I would add a simple example in the coming days. Currently the typescript sdk example folder is called nodejs. My approach would be to rename it to typescript and add a bun example on the same level as react-build. Any additional input on what the example should include is also welcome.

I don't think we need an example in our directory, it's pretty much outdated.

Instead, we should write a page in the TypeScript developer manual: https://docs.dagger.io/manuals/developer/typescript on how to use Bun (@vikram-dagger is the owner of this part).

We can keep it short, just explaining the `package.json` and how to execute bun / explain the gain of performance and give a warning to the user because it's a new feature that may be unstable (we tested it but since Bun isn't fully compatible with Node for now, people might hit issues)

**Comment by vikram-dagger** [comment]
> > I would add a simple example in the coming days. 
> 
> Instead, we should write a page in the TypeScript developer manual: https://docs.dagger.io/manuals/developer/typescript on how to use Bun (@vikram-dagger is the owner of this part).

The TypeScript developer manual is the correct location for this example. I would add a page, perhaps called "Alternative Runtimes"? It could have a section for Bun, including a code example. This would allow us to extend the same page for other runtimes if such support is added in future. 

If this is agreeable, I can add a page but will need your help @TomChv to fill it out with the required information, as I'm not very familiar with this topic.

**Comment by benjaminstrasser** [comment]
> @benjaminstrasser wants to do it, I was also planing to take care of that so no worry, we'll just need you for a review haha
> 
> @benjaminstrasser I can open a draft PR and we can co-work on it if you want?
@TomChv Sounds good to me. Just ping me once the PR is open.

**Comment by TomChv** [comment]
@benjaminstrasser @vikram-dagger I created the PR https://github.com/dagger/dagger/pull/7054
I'm waiting for your comments 🚀

**Comment by benjaminstrasser** [comment]
@TomChv 
To me it looks like we can close this issue.
Any open tasks we forgot about?

**Comment by TomChv** [comment]
Nope, everything looks good to me!

Thanks again for your amazing contribution! 🚀 
There's so more job to do related to optimize the TypeScript SDK itself but it's not related to Bun directly.
